jena-arq/src/main/java/org/apache/jena/sparql/core/mem/DatasetGraphInMemory.java

```diff
@@ -161,17 +161,17 @@ public void abort() {
 	@Override
 	public void close() {
 		if (isInTransaction()) abort();
-
 	}
 
 	@Override
 	public void end() {
         if (isInTransaction()) {
+            log.warn("Ending transaction without commit!");
```

Comment:
begin(READ)-end() is a normal pattern.  The code needs to know if it is considering a READ or WRITE transaction.